src/block-cache.ts

```diff
@@ -0,0 +1,183 @@
+import {
+  createAsyncMiddleware,
+  JsonRpcMiddleware,
+  JsonRpcRequest,
+} from 'json-rpc-engine';
+import * as cacheUtils from './cache-utils';
+
+// eslint-disable-next-line @typescript-eslint/no-var-requires,@typescript-eslint/no-require-imports
+const BlockTracker = require('eth-block-tracker');
+
+// `<nil>` comes from https://github.com/ethereum/go-ethereum/issues/16925
+const emptyValues: (string|null|undefined)[] = [undefined, null, '\u003cnil\u003e'];
+
+interface BlockCacheMiddlewareOptions{
+  blockTracker?: typeof BlockTracker;
+}
+
+export = createBlockCacheMiddleware;
+
+//
+// Cache Strategies
+//
+
+class BlockCacheStrategy {
+
+  private cache: Record<string, Record<string, unknown>>;
+
+  constructor() {
+    this.cache = {};
+  }
+
+  getBlockCacheForPayload(_payload: JsonRpcRequest<string[]>, blockNumberHex: string): Record<string, unknown> {
+    const blockNumber: number = Number.parseInt(blockNumberHex, 16);
+    let blockCache: Record<string, unknown> = this.cache[blockNumber];
+    // create new cache if necesary
+    if (!blockCache) {
+      const newCache: Record<string, unknown> = {};
+      this.cache[blockNumber] = newCache;
+      blockCache = newCache;
+    }
+    return blockCache;
+  }
+
+  async get(payload: JsonRpcRequest<string[]>, requestedBlockNumber: string): Promise<unknown> {
```

Comment:
I think we can do better here than `Promise<unknown>`. We shouldn't have to cast the return value of this function where it's used.